app/abci/handlers.go

```diff
@@ -0,0 +1,270 @@
+		logger:                 logger,
+	}
+}
+
+func (h *Handlers) SetSEDASigner(signer utils.SEDASigner) {
+	h.signer = signer
+}
+
+// ExtendVoteHandler handles the ExtendVote ABCI to sign a batch created
+// from the previous block.
+func (h *Handlers) ExtendVoteHandler() sdk.ExtendVoteHandler {
+	return func(ctx sdk.Context, _ *abci.RequestExtendVote) (*abci.ResponseExtendVote, error) {
+		h.logger.Debug("start extend vote handler")
+
+		// Sign the batch created from the last block's end blocker.
+		batch, err := h.batchingKeeper.GetBatchForHeight(ctx, ctx.BlockHeight()+voteExtensionOffset)
+		if err != nil {
+			return nil, err
+		}
+		signature, err := h.signer.Sign(batch.BatchId, utils.Secp256k1)
+		if err != nil {
+			return nil, err
+		}
+
+		h.logger.Debug("submitting batch signature", "signature", signature)
+		return &abci.ResponseExtendVote{VoteExtension: signature}, nil
+	}
+}
+
+// VerifyVoteExtensionHandler handles the VerifyVoteExtension ABCI to
+// verify the batch signature included in the pre-commit vote against
+// the public key registered in the pubkey module.
+func (h *Handlers) VerifyVoteExtensionHandler() sdk.VerifyVoteExtensionHandler {
+	return func(ctx sdk.Context, req *abci.RequestVerifyVoteExtension) (*abci.ResponseVerifyVoteExtension, error) {
+		h.logger.Debug("start verify vote extension handler", "request", req)
+
+		if len(req.VoteExtension) != 64 {
+			h.logger.Error("invalid vote extension length", "len", len(req.VoteExtension))
+			return nil, ErrInvalidVoteExtensionLength
+		}
+
+		batch, err := h.batchingKeeper.GetBatchForHeight(ctx, ctx.BlockHeight()+voteExtensionOffset)
+		if err != nil {
+			return nil, err
+		}
+		ok, err := h.verifyBatchSignature(ctx, batch.BatchId, req.VoteExtension[:64], req.ValidatorAddress, utils.Secp256k1)
+		if !ok || err != nil {
+			return nil, err
+		}
+
+		h.logger.Debug("successfully verified signature", "request", req.ValidatorAddress, "height", req.Height)
+		return &abci.ResponseVerifyVoteExtension{Status: abci.ResponseVerifyVoteExtension_ACCEPT}, nil
+	}
+}
+
+// PrepareProposalHandler handles the PrepareProposal ABCI to inject
+// a canonical set of vote extensions in the proposal.
+func (h *Handlers) PrepareProposalHandler() sdk.PrepareProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+		h.logger.Debug("start prepare proposal handler")
+
+		var injection []byte
+		if req.Height > ctx.ConsensusParams().Abci.VoteExtensionsEnableHeight {
+			err := baseapp.ValidateVoteExtensions(ctx, h.stakingKeeper, req.Height, ctx.ChainID(), req.LocalLastCommit)
+			if err != nil {
+				return nil, err
+			}
+
+			injection, err = json.Marshal(req.LocalLastCommit)
+			if err != nil {
+				h.logger.Error("failed to marshal extended votes", "err", err)
+			}
+
+			injectionSize := int64(len(injection))
+			if injectionSize <= req.MaxTxBytes {
+				req.MaxTxBytes -= injectionSize
+			} else {
```

Comment:
Personal preference but we can also use an early return here to drop the `else` clause :)